File: src/lib/model/activities/activity.ts
import IActivity from "./activity-interface";

class Activity implements IActivity {
  private _type: string;
  private _sectionNum: string;
  private _daysOfWeek: string;
  private _timeOfDay: string;
  private _building: string;
  private _room: string;

  constructor(
    type: string,
    sectionNum: string,
    daysOfWeek: string,
    timeOfDay: string,
    building: string,
    room: string,
  ) {
    this._type = type;
    this._sectionNum = sectionNum;
    this._daysOfWeek = daysOfWeek;
    this._timeOfDay = timeOfDay;
    this._building = building;
    this._room = room;
  }

  get type(): string {
    return this._type;
  }

  get sectionNum(): string {
    return this._sectionNum;
  }

  get daysOfWeek(): string {
    return this._daysOfWeek;
  }

  get timeOfDay(): string {
    return this._timeOfDay;
  }

  get building(): string {
    return this._building;
  }

  get room(): string {
    return this._room;
  }

  get location(): string {
    return `${this._building} ${this._room}`;
  }
}

export default Activity;

File: scripts/models/activity.py
class Activity:
    def __init__(
        self,
        section_id: int,
        meeting_type: str,
        section_num: str,
        days_of_week: str,
        time: str,
        building: str,
        room: str,
        instructor: str,
        id: int = -1,
    ):
        self._section_id = section_id
        self._meeting_type = meeting_type
        self._section_num = section_num
        self._days_of_week = days_of_week
        self._time = time
        self._building = building
        self._room = room
        self._instructor = instructor
        self._id = id

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Activity):
            return False

        return (
            self._section_id == other._section_id
            and self._meeting_type == other._meeting_type
            and self._section_num == other._section_num
            and self._days_of_week == other._days_of_week
            and self._time == other._time
            and self._building == other._building
            and self._room == other._room
            and self._instructor == other._instructor
        )

    def __hash__(self) -> int:
        return hash(
            self._section_id,
            self._meeting_type,
            self._section_num,
            self._days_of_week,
            self._time,
            self._building,
            self._room,
            self._instructor,
        )

    def get_section_id(self) -> int:
        return self._section_id

    def get_meeting_type(self) -> str:
        return self._meeting_type

    def get_section_num(self) -> str:
        return self._section_num

    def get_days_of_week(self) -> str:
        return self._days_of_week

    def get_time(self) -> str:
        return self._time

    def get_building(self) -> str:
        return self._building

    def get_room(self) -> str:
        return self._room

    def get_instructor(self) -> str:
        return self._instructor

    def get_id(self) -> int:
        return self._id

    def with_id(self, id: int) -> "Activity":
        return Activity(
            self._section_id,
            self._meeting_type,
            self._section_num,
            self._days_of_week,
            self._time,
            self._building,
            self._room,
            self._instructor,
            id,
        )
